modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java

```diff
@@ -1197,44 +1202,39 @@ private CompletableFuture<Void> remoteProcessingFuture(DisasterRecoveryRequest r
 
         MultiNodeDisasterRecoveryRequest multiNodeRequest = (MultiNodeDisasterRecoveryRequest) request;
 
-        Collection<String> actualNodeNames = getActualNodeNames(multiNodeRequest.nodeNames());
+        Set<NodeWithAttributes> nodes = getRequestNodes(multiNodeRequest.nodeNames());
 
-        CompletableFuture<?>[] remoteProcessingFutures = actualNodeNames
+        CompletableFuture<?>[] remoteProcessingFutures = nodes
                 .stream()
-                .map(nodeName -> addMultiNodeOperation(nodeName, operationId))
+                .map(node -> addMultiNodeOperation(node, operationId))
                 .toArray(CompletableFuture[]::new);
 
-        return allOf(remoteProcessingFutures)
-                .whenComplete((ignored, e) -> {
-                    for (String nodeName : actualNodeNames) {
-                        operationsByNodeName.compute(nodeName, (node, operations) -> {
-                            if (operations != null) {
-                                operations.remove(operationId);
-
-                                return operations.isEmpty() ? null : operations;
-                            }
-
-                            return null;
-                        });
-                    }
-                });
+        return allOf(remoteProcessingFutures);
     }
 
     /** If request node names is empty, returns all nodes in the logical topology. */
-    private Collection<String> getActualNodeNames(Set<String> requestNodeNames) {
-        if (requestNodeNames.isEmpty()) {
-            return dzManager.logicalTopology().stream()
-                    .map(NodeWithAttributes::nodeName)
-                    .collect(toSet());
-        } else {
-            return requestNodeNames;
-        }
+    private Set<NodeWithAttributes> getRequestNodes(Set<String> requestNodeNames) {
+        return dzManager.logicalTopology().stream()
+                .filter(node -> requestNodeNames.isEmpty() || requestNodeNames.contains(node.nodeName()))
+                .collect(toSet());
     }
 
-    private CompletableFuture<Void> addMultiNodeOperation(String nodeName, UUID operationId) {
-        CompletableFuture<Void> result = new CompletableFuture<Void>().orTimeout(TIMEOUT_SECONDS, TimeUnit.SECONDS);
+    private CompletableFuture<Void> addMultiNodeOperation(NodeWithAttributes node, UUID operationId) {
+        CompletableFuture<Void> result = new CompletableFuture<Void>().orTimeout(DISASTER_RECOVERY_TIMEOUT_MILLIS, TimeUnit.SECONDS);
+
+        operationsByNodeId.compute(node.nodeId(), (nodeId, operations) -> {
+            Set<UUID> nodes = dzManager.logicalTopology().stream()
+                    .map(NodeWithAttributes::nodeId)
+                    .collect(toSet());
+
+            if (!nodes.contains(nodeId)) {
+                result.completeExceptionally(new IllegalStateException(
```

Comment:
Should it be completed with the same exception class you use on 'node left' handling?